compiler/src/iree/compiler/API2/Embed.h

```diff
@@ -0,0 +1,257 @@
+// legal to destroy nullptr.
+IREE_EMBED_EXPORTED void ireeCompilerErrorDestroy(
+    struct iree_compiler_error_t *error);
+
+// Gets the message associated with the error as a C-string. The string will be
+// valid until the error is destroyed.
+IREE_EMBED_EXPORTED const char *ireeCompilerErrorGetMessage(
+    struct iree_compiler_error_t *error);
+
+//===----------------------------------------------------------------------===//
+// Global initialization.
+//===----------------------------------------------------------------------===//
+
+// Gets the version of the API that this compiler exports.
+IREE_EMBED_EXPORTED int ireeCompilerGetAPIVersion();
+
+// The compiler must be globally initialized before further use.
+// It is intended to be called as part of the hosting process's startup
+// sequence. Any failures that it can encounter are fatal and will abort
+// the process.
+//
+// If |initializeCommandLine| is true, then compiler flags will be bound to
+// the process-level command line environment. This can provide convenient
+// mnemonic access to single tenant setups but is risky in a multi-tenant
+// library.
+//
+// Note that for internal tools which use InitLLVM to manage access to the CLI,
+// it is legal for that to co-exist with this global initialize/shutdown.
+IREE_EMBED_EXPORTED void ireeCompilerGlobalInitialize(
+    bool initializeCommandLine);
+
+// Destroys any process level resources that the compiler may have created.
+// This must be called prior to library unloading.
+IREE_EMBED_EXPORTED void ireeCompilerGlobalShutdown();
+
+// Invokes a callback with each registered HAL target backend.
+IREE_EMBED_EXPORTED void ireeCompilerEnumerateRegisteredHALTargetBackends(
+    void (*callback)(const char *backend, void *userData), void *userData);
+
+//===----------------------------------------------------------------------===//
+// Session management.
+// A session represents a scope where one or more runs can be executed.
+// Internally, it consists of an MLIRContext and a private set of session
+// options. If the CL environment was initialized, session options will be
+// bootstrapped from global flags.
+//
+// Session creation cannot fail in a non-fatal way.
+//===----------------------------------------------------------------------===//
+
+// Creates a new session (which must be destroyed by
+// ireeCompilerSessionDestroy).
+IREE_EMBED_EXPORTED struct iree_compiler_session_t *ireeCompilerSessionCreate();
+
+// Destroys a session.
+IREE_EMBED_EXPORTED void ireeCompilerSessionDestroy(
+    struct iree_compiler_session_t *session);
+
+//===----------------------------------------------------------------------===//
+// Run management.
+// Runs execute against a session and represent a discrete invocation of the
+// compiler.
+//===----------------------------------------------------------------------===//
+
+IREE_EMBED_EXPORTED struct iree_compiler_run_t *ireeCompilerRunCreate(
```

Comment:
naming nit: maybe "invocation" instead of "run"? (run is somewhat ambiguous as verb + noun and for some reason doesn't recognize as repeatable to me - maybe because usually I see people talking about "invoking APIs" instead of "running" them like standalone tools)